sycl/include/CL/sycl/detail/image_accessor_util.hpp

```diff
@@ -123,85 +188,533 @@ template <typename T> vec<T, 4> readPixel(T *Ptr, image_channel_order Order) {
-  RetData.y() = (float)PixelData.y();
-  RetData.z() = (float)PixelData.z();
-  RetData.w() = (float)PixelData.w();
+
+  switch (ImageChannelType) {
+  case image_channel_type::snorm_int8:
+  case image_channel_type::snorm_int16:
+  case image_channel_type::unorm_int8:
+  case image_channel_type::unorm_int16:
+  case image_channel_type::unorm_short_565:
+  case image_channel_type::unorm_short_555:
+  case image_channel_type::unorm_int_101010:
+    // TODO: Missing information in OpenCL spec.
+    throw cl::sycl::feature_not_supported(
+        "Currently unsupported datatype conversion from image_channel_type "
+        "to cl_half4.");
+    break;
+  case image_channel_type::signed_int8:
+  case image_channel_type::signed_int16:
+  case image_channel_type::signed_int32:
+  case image_channel_type::unsigned_int8:
+  case image_channel_type::unsigned_int16:
+  case image_channel_type::unsigned_int32:
+    // OpenCL Spec section 6.12.14.2 does not allow reading float4 data to an
+    // image with channel datatype - signed/unsigned_int8,signed/unsigned_int16
+    // and signed/unsigned_int32.
+    throw cl::sycl::invalid_parameter_error(
+        "Datatype to read- cl_half4 is incompatible with the "
+        "image_channel_type of the image.");
+    break;
+  case image_channel_type::fp16:
+    RetData = PixelData.template convert<cl_half>();
+    break;
+  case image_channel_type::fp32:
+    throw cl::sycl::invalid_parameter_error(
+        "Datatype to read - cl_half4 is incompatible with the "
+        "image_channel_type of the image.");
+    break;
+  default:
+    break;
+  }
+  return PixelData;
 }
 
 // Converts data to write into appropriate datatype based on the channel of the
 // image.
-// TODO: Change this method to use the conversion rules as given in the OpenCL
-// Spec Section 8.3. The conversion rules may be handled differently for each
-// return datatype - float, int32, uint32, half. ImageChannelType is passed to
-// the function to use appropriate conversion rules.
-template <typename ChannelType, typename WriteDataType>
-vec<ChannelType, 4> convertWriteData(vec<WriteDataType, 4> WriteData,
+// The conversion rules used are as given in OpenCL Spec Section 8.3. The
+// conversion rules are different for each return datatype - float,
+// int32, uint32, half. ImageChannelType is passed to the function to use
+// appropriate conversion rules.
+template <typename ChannelType>
+vec<ChannelType, 4> convertWriteData(vec<cl_uint, 4> WriteData,
                                      image_channel_type ImageChannelType) {
   vec<ChannelType, 4> PixelData;
-  PixelData.x() = (ChannelType)WriteData.x();
-  PixelData.y() = (ChannelType)WriteData.y();
-  PixelData.z() = (ChannelType)WriteData.z();
-  PixelData.w() = (ChannelType)WriteData.w();
+  vec<cl_uint, 4> Temp;
+  cl_uint MinVal;
+  cl_uint MaxVal;
+  switch (ImageChannelType) {
+  case image_channel_type::unsigned_int8:
+    // convert_uchar_sat(Data)
+    MinVal = min_v<cl_uchar>();
```

Comment:
```suggestion
    const cl_uint MinVal = min_v<cl_uchar>();
```
in other places do the same